### DISCUSSION TITLE
"onto" - TTS sequence

### DISCUSSION THREAD

**Discussion by ZeljkoBG** [organicmaps/organicmaps#9702] [Translations]
sequence "onto", OM used in two different cases.

**Скрените десно** "onto" **Паунова** (street name)
and the second case when it refers to "direction".

All this would be great if the car was driven by Tarzan.
In Slavic languages, this is not possible because they are complicated (grammar, падежи, ...)

Poland, Slovakia, Serbia,... do not have the noun 'STREET' either before or after the name of the street.
So the sequence is grammatically incorrect.

Turn right / onto / **Паунова** (nominative)
for it to be correct it is needed (accusative) **Паунову**
I CANNOT use the accusative case. I have to use the nominative - **Паунова**

To get the correct sentence, I translate 'onto'  - 'to the street'. So I get the correct sentence:
Скрените десно/ у улицу / Паунова

**The SECOND problem** arises because you use 'onto' for direction as well. 
Turn right/ onto / Center.
**The sentence should be longer** and that it reads like this

Turn right /onto/ **Паунова**. Direction Center/ highway/airport/...
Скрените десно/ у улицу / **Паунова**. Правац Центар  / аутопут/аеродром/...

You need to add TTS tag for DIRECTION. Like the code.
Like OsmAnd, iGO,...
dictionary["toward"] = tts ? "Правац " : "toward.ogg";

PS:
tomorrow is my family's religious holiday and we have guests coming. If I devoted the evening to Organic Maps, then I deserved to get an answer:
Did you understand the problem?
Will you correct it?

**Comment by biodranik** [comment]
@zyphlar can you please help?

**Comment by zyphlar** [comment]
@ZeljkoBG thank you for this feedback! Would you be willing to join our Bulgarian translations team so you're aware of issues with Bulgarian and able to help? Right now I'm not sure we have any native Bulgarian speakers, so that will greatly help: https://github.com/orgs/organicmaps/teams/translations-bg

Secondly we currently experience this issue in most languages: for example in English it will currently say "in 500 meters, exit onto 345, London, City Center, Main Street" even though the most proper format would be more like "in 500 meters, take exit 345 onto Main Street for City Center towards London". This is because as you can imagine the complexity for translations increases hugely as we add more variables for words to be inserted, especially in languages with complex grammar rules.

Could you provide translations for some of the following phrases, so that the next revision of the Street Name TTS feature can make better attempts? Feel free to replace names with a common name found in your country. If there are complex situations like nominative/accusative, feel free to provide multiple translations, one for each case. (We have a lot of logic necessary for Hungarian, for example, because the verbs need to change based on whether the noun has high vs low sounds.)

Feel free to comment if some of the phrases have special cases or are unclear, etc.

Thank you again for your help!

- EN: In 500 meters, turn right onto Main Street.
  - BG: 
- EN: In 500 meters, turn right onto Highway 123.
  - BG: 
- EN: In 500 meters, exit left onto Main Street. (or "take the left exit", for motorways with left-hand-side exits)
  - BG: 
- EN: In 500 meters, exit right onto Main Street. (or "take the right exit", for motorways, when the exit sign only has a street name)
  - BG: 
- EN: In 500 meters, exit onto Main Street. (or "take the exit", for motorways, when the exit side is unclear)
  - BG: 
- EN: In 500 meters, take exit 345.
  - BG: 
- EN: In 500 meters, take exit 345 onto Main Street.
  - BG: 
- EN: In 500 meters, take exit 345 for City Center.
  - BG: 
- EN: In 500 meters, take exit 345 towards London.
  - BG: 
- EN: In 500 meters, take exit 345 for City Center towards London.
  - BG: 
- EN: In 500 meters, take exit 345 onto Main Street for City Center towards London.
  - BG: 
- EN: In 500 meters, take exit 345 onto Highway 123 for City Center towards London.
  - BG: 
- EN: In 500 meters, take the exit for City Center.
  - BG: 
- EN: In 500 meters, take the exit towards London.
  - BG: 
- EN: In 500 meters, take the exit for City Center towards London.
  - BG: 
- EN: In 500 meters, take the left-hand exit 345 onto Main Street for City Center towards London.
  - BG:

**Comment by zyphlar** [comment]
cc #8412

**Reply by ZeljkoBG** [reply]
Whatever you need, I'm here to help. But I'm not Bulgarian! I am a Serb.

ZeljkoBG
BG --> Belgrade
:)

**Reply by ZeljkoBG** [reply]
1. In Bulgarian, there is no STREET in the name. Only the name "Неофит Рилски". 

2. I would bet that it is the same as in the Serbian language. But I don't know if you understood my post.
3. The solution is simple, I suggested it.

Your example is unclear to me
"EN: In 500 meters, take exit 345 towards London."
Are you entering the city or have you just left for London?

"EN: In 500 meters, take exit 345 for City Center towards London."
It's simpler this way - In 500 meters, take exit 345. Towards Center.
If you're entering a city, it's logical that you know which one it is. But it is not logical for you to know where the center is. That's why "Towards Center."


Note (valid only from Serbia): no citizen knows the number of exits. For us, this is irrelevant and boring information.
Please see how OsmAnd did TTS for each language. Excellent solution, except for 'exit number' of course.

**Reply by zyphlar** [reply]
Ah! I see you're already on our Serbian language team. Excellent, and thanks!

I understand that many streets don't use the word "street", that's okay. It would be odd to say "Turn right onto Peacock" in English so that's why I say `Main Street` as a name, but any other name is fine. We could use "Turn right onto `Broadway`" as an example. Or Паунова. Your choice :) Organic Maps will probably not _remove_ partial contents of the Street Name variable, so if a Slavic user drives in an English speaking country the TTS may still say something like "Скрените десно у улицу Main Street" but that's obviously not a concern for driving in a country with the same main language as the app. (In OpenStreetMap we write whatever is displayed on the signs, so it would be odd to be presumptuous and i.e. translate a proper noun like `Main Street` into `Main 通り` for Japanese users driving in America.)

"EN: In 500 meters, take exit 345 towards London" would likely be a sign that just says "`345` London 👉" meaning that you will be going in the direction of London if you take this exit (and not if you don't). Many countries will put the names of cities or places as example destinations on exit signs so you can navigate between roads, cities, or countries without needing a map. "This exit/road will take you to London" (from somewhere not-London. Whether we're imminently entering London or having just left our home on a long journey towards London is not specified, we're just reading what the sign says. "In the direction of.")

As for `City Center` that's one example of a destination in my area. It could be anything, like `History Museum` or `Beaches` or `Main Hospital`. Organic Maps doesn't get to choose what these words are, again we're just reading the sign. It may help to click on some of the [images on this page](https://zyphlar.github.io/organicmaps-locale-viewer/) as they're examples of exit signs that Organic Maps might need to read. If the sign says "London," Organic Maps needs to read "London" and the only question is how to grammatically do it. (There are some exits or motorway splits in my area that have no numbers or names, only cities! "Exit left towards San Francisco, or Exit right towards Los Angeles". We have to handle every case even if it's rare in our region. You can see why I released this feature with all the exit identifiers mashed together with "onto" rather than trying to figure this out for two dozen languages I don't speak 😉)

It's okay to ignore the exit number examples, however if you can provide them to be grammatically correct it would be nice (Serbian speakers may travel to an area which does have exit numbers, and we don't want them to be confused!)

I can't simply copy OsmAnd unfortunately, and anyway we want to ultimately do a better job than them. In my region OsmAnd's voice directions are actually worse than Organic Maps! We have a goal of being natural and understandable to everyone in all countries.

It may be helpful for you to think of OSM key names (variables) instead of actual human language, since that's what we ultimately need to write in C++ anyway. [See this OSM wiki page for details and examples.](https://wiki.openstreetmap.org/wiki/Tag:highway%3Dmotorway_junction) We have:

- Destination Ref (The new motorway(s) we will be on if we take this exit, like A500 or I-50 or Highway 23 or "A31, A39". Might be used with "onto" or "via")
- Junction Ref (The identifier of this exit)
- Destination (The names of one or more cities or points of interest we will be headed towards if we take this exit, like "City Center, London")
- Road Name (The non-motorway road name we will be on if we take this exit, like "Main Street" or "Паунова")

Thank you for your help!

**Comment by ZeljkoBG** [comment]
To summarize and simplify my proposal.

1. Add "Direction" to TTS text. Just one word "Direction". Of course it
only refers to Direction instead of the incorrect 'onto'. Then the sentence
will be grammatically correct in all languages.

2. For English and similar languages that have the word 'Street' on the map
"onto":" onto",
//  onto street ":" onto street",   ---> inactive

3. For languages, Serbian, Bulgarian, Polish,... that do not have 'Street' on
// "onto":"onto ",   ---> inactive
"onto":"onto street",    ---> where we will translate grammatically
correctly and read the name of the street in the nominative case   "Paunova"

2a.  Turn right  "onto" Studley Road. "Direction" airport
2b. Skreni desno "u ulicu" Paunova. "Pravac" aerodrom

**Reply by zyphlar** [reply]
Thank you! Most of this is not advice I can simply implement unfortunately. The next version of the TTS will be much more complex in order to handle more languages and cases more properly, so adding "simple" strings might be a temporary fix but doesn't help the overall situation. It needs to be rewritten, probably with the help of a more robust localization library.

The best help I could get right now would be if you were able to provide the best most grammatically friendly translations of these strings. That way when I redesign the TTS I can create tests and ensure proper results:


    EN: In 500 meters, turn right onto Peacock.
        SR:
    EN: In 500 meters, turn right onto Highway 123.
        SR:
    EN: In 500 meters, exit left onto Peacock. (or "take the left exit", for motorways with left-hand-side exits)
        SR:
    EN: In 500 meters, exit right onto Peacock. (or "take the right exit", for motorways with right-hand-side exits)
        SR:
    EN: In 500 meters, exit onto Peacock. (or "take the exit", for motorways, when the exit side is unclear)
        SR:
    EN: In 500 meters, take exit 345.
        SR:
    EN: In 500 meters, take exit 345 onto Peacock.
        SR:
    EN: In 500 meters, take exit 345 for Airport.
        SR:
    EN: In 500 meters, take exit 345 towards London.
        SR:
    EN: In 500 meters, take exit 345 for Airport towards London.
        SR:
    EN: In 500 meters, take exit 345 onto Peacock for Airport towards London.
        SR:
    EN: In 500 meters, take exit 345 onto Highway 123 for Airport towards London.
        SR:
    EN: In 500 meters, take the exit for Airport.
        SR:
    EN: In 500 meters, take the exit towards London.
        SR:
    EN: In 500 meters, take the exit for Airport towards London.
        SR:
    EN: In 500 meters, take the left-hand exit 345 onto Peacock for Airport towards London.
        SR:

**Comment by ZeljkoBG** [comment]
EN: In 500 meters, turn right onto Peacock.
    SR: За 500 метара, скренуте десно у улицу Peacock
EN: In 500 meters, turn right onto Highway 123.
    SR: За 500 метара, скренуте десно на аутопут 123.
EN: In 500 meters, exit left onto Peacock. (or "take the left exit", for motorways with left-hand-side exits)
    SR: За 500 метара, изађите лево у улицу Peacock.
EN: In 500 meters, exit right onto Peacock. (or "take the right exit", for motorways with right-hand-side exits)
    SR: За 500 метара, изађите десно у улицу Peacock.
EN: In 500 meters, exit onto Peacock. (or "take the exit", for motorways, when the exit side is unclear)
    SR: За 500 метара, изађите у улицу Peacock.
EN: In 500 meters, take exit 345.
    SR: За 500 метара, идите на излаз 345.
EN: In 500 meters, take exit 345 onto Peacock.
    SR: За 500 метара, идите на излаз 345 у улицу Peacock.
EN: In 500 meters, take exit 345 for Airport.
    SR: За 500 метара, идите на излаз 345 за аеродром.
EN: In 500 meters, take exit 345 towards London.
    SR: За 500 метара, идите на излаз 345 за London.
EN: In 500 meters, take exit 345 for Airport towards London.
    SR: За 500 метара, идите на излаз 345 за аеродром. Правац London.
EN: In 500 meters, take exit 345 onto Peacock for Airport towards London.
    SR: За 500 метара, идите на излаз 345 у улицу Peacock за аеродром. Правац London.
EN: In 500 meters, take exit 345 onto Highway 123 for Airport towards London.
    SR: За 500 метара, идите на излаз 345 на аутопут 123 за аеродром. Правац London.
EN: In 500 meters, take the exit for Airport.
    SR: За 500 метара, идите на излаз за аеродром.
EN: In 500 meters, take the exit towards London.
    SR: За 500 метара, идите на излаз. Правац London
EN: In 500 meters, take the exit for Airport towards London.
    SR: За 500 метара, идите на излаз за аеродром. Правац London.
EN: In 500 meters, take the left-hand exit 345 onto Peacock for Airport towards London.
    SR: За 500 метара, идите на леви излаз 345 у улицу Peacock за аеродром. Правац London.

**Reply by zyphlar** [reply]
Thank you so much! When it's time to redo the TTS this will be very helpful in making the grammar better.

**Reply by ZeljkoBG** [reply]
Don't thank me.
I'm here to help, and give advice, to make Organic a better navigation.

**Comment by ZeljkoBG** [comment]
@zyphlar there are some flaws i noticed in your template.

There is a maneuver:
turn right  -  for street   ---> In 500 meters, turn right onto Peacock.
 
there is no maneuver
**turn left**  -  for street

or i don't understand?

**Reply by zyphlar** [reply]
That's okay! Is the grammar very different? Feel free to provide "left" here just for completeness. There are probably a hundred possible combinations including roundabouts so I just included the most important variations relevant to "onto"

**Comment by ZeljkoBG** [comment]
EN: In 500 meters, turn left onto Peacock.
SR: За 500 метара, скрените лево у улицу Peacock

Anything you need, let me know.
When the new TTS text is implemented, I will do it for the Croatian language as well.

**Comment by ZeljkoBG** [comment]
@zyphlar
example instruction:

"Turn right" = "Скрените десно"

"Turn right" / "then" / in 400 meters / "Turn right"
BUT
Скрените десно / затим/ за 400 метара / Скре**ћете** десно

"**Скрећете десно**" is future tense. We will only perform that action later.

**Reply by zyphlar** [reply]
Thanks, can you provide future tenses for all the instructions in sound.txt?

**Reply by ZeljkoBG** [reply]
You know that best. The only clear line is:
"destination":"Стижеш на одредиште.",

If I'm not mistaken, all verbs/actions that have a tense in the sentence must have "future tenses"
As a previous example where the action will be performed in the future tense.

example:
 *= "future tenses"

... / THEN / in 400 meters / *
... / THEN / in 500 meters / *
ETC

**Comment by ZeljkoBG** [comment]
@zyphlar @biodranik 

**Off topic** - urgently
 
I translated the application into Serbian Cyrillic. I don't know how to post a translation myself. Please add it so that the translation can be found in the new version.

[strings.txt](https://github.com/user-attachments/files/17948072/strings.txt)

**Comment by ZeljkoBG** [comment]
I see I've been added. Thanks.
I don't see
https://github.com/organicmaps/organicmaps/blob/master/data/strings/strings.txt that the translation has been added.
Is the 'strings.txt' file translation I did in vain?

**Reply by zyphlar** [reply]
Not in vain we need a programmer to add Serbian support

**Reply by ZeljkoBG** [reply]
You're worth half the application. If it wasn't for you, I would have given up.

**Reply by zyphlar** [reply]
@ZeljkoBG haha you might have me confused for someone else, my major contribution is only adding street names to TTS ;)

I chatted with @biodranik and @vng and they say you'll need to get Serbian translations for more files than just strings.txt before we can enable it as a language option (so the app looks mostly translated and not weirdly half English) -- this can be a significant project so maybe you can enlist help with some Serbian friends? The full list of files is here, plus general translation instructions:

https://github.com/organicmaps/organicmaps/blob/master/docs/TRANSLATIONS.md#translation-files

Then you or we can create a Pull Request for these file changes. If you don't want to make a Pull Request, it may be a good idea to at least create a new Discussion or Issue for Serbian, to hold these files separately from this "onto" issue since they're two different projects.

**Reply by ZeljkoBG** [reply]
I don't know who is responsible for what. I put the translation here because I don't know where to put it. I had a similar experience 5-6 months ago. I made a translation and I didn't understand what I was supposed to do. They told me something like '*.TXT you import and TXT is automatically converted to *.JSON...!
 
I still don't understand what I need to translate! What are the files. Although I wrote here that I 'want to translate' and get a link https://github.com/organicmaps/organicmaps/blob/master/docs/TRANSLATIONS.md#translation-files
 and now what?

this?
Main:

Application UI strings: [data/strings/strings.txt](https://github.com/organicmaps/organicmaps/blob/master/data/strings/strings.txt)
A few iOS specific strings: [iphone/plist.txt](https://github.com/organicmaps/organicmaps/blob/master/iphone/plist.txt)
Names of map features/types: [data/strings/types_strings.txt](https://github.com/organicmaps/organicmaps/blob/master/data/strings/types_strings.txt)
Search keywords/aliases/synonyms for map features: [data/categories.txt](https://github.com/organicmaps/organicmaps/blob/master/data/categories.txt)


Or this must too
Additional:

Text-to-speech strings for navigation: [data/strings/sound.txt](https://github.com/organicmaps/organicmaps/blob/master/data/strings/sound.txt)

Android stores description: [android/app/src/fdroid/play/](https://github.com/organicmaps/organicmaps/blob/master/android/app/src/fdroid/play)

Apple AppStore description: [iphone/metadata/](https://github.com/organicmaps/organicmaps/blob/master/iphone/metadata)

Popular brands of map features: [data/strings/brands_strings.txt](https://github.com/organicmaps/organicmaps/blob/master/data/strings/brands_strings.txt)

Search keywords for popular brands: [data/categories_brands.txt](https://github.com/organicmaps/organicmaps/blob/master/data/categories_brands.txt)

Search keywords for cuisine types: [data/categories_cuisines.txt](https://github.com/organicmaps/organicmaps/blob/master/data/categories_cuisines.txt)

Country / map region names: [data/countries_names.txt](https://github.com/organicmaps/organicmaps/blob/master/data/countries_names.txt)

[other strings](https://github.com/organicmaps/organicmaps/blob/master/docs/STRUCTURE.md#strings-and-translations) files

Or is it written somewhere? When no one addressed me with words, you should translate: this and that file and everything clearly.


I will not interfere in your business politics. Do as you like. Just to help you out, it doesn't make any sense. Look at any app: what's translated, is translated. The new untranslated parts are in English. Only you are in dispute. 

To translate into Cyrillic "popular brands"? - Coca Cola,... that doesn't translate into Serbian!
To translate into Cyrillic "map region names"? - "Brest, Quimper, Concarneau", "Macau, Djibouti and Maruti", I don't have the time or desire for that.
To translate into Cyrillic "iOS specific strings"? -  I'm not interested in that, and I don't have iOS, so I can't check in the application whether the translation is correct.
...


PS_1: I translated the application and that's it for now. If it suits you, please add a translation. If it doesn't suit you, then nothing.
PS_2: I'm staying to make a new TTS text. Report everything you need.

**Reply by zyphlar** [reply]
@ZeljkoBG that's okay, you don't need to translate every single thing especially right away. I think the best thing to do for now is if you make a new issue called Serbian language support and upload your strings.txt there.

Since Serbian is a Cyrillic language we will probably want to translate most strings like store, address, pizza, etc, so that it doesn't look like an "English app" to people in Serbia, but it can be a lot of work and we can do it piece by piece especially if you know any Serbians who might like to help.

**Reply by ZeljkoBG** [reply]
Of course. To go gradually. Slavic languages ​​are complicated. I have to look in the app and fix the errors.
Not a single word will be in English. It is normal for the terms 'store, address, pizza, etc' to be translated. There is no discussion about that. And don't worry. If I came forward to do it, it will be done, but the states will not, for now. My translations are used by Sygic, Navitel, Sygic Truck, OsmAnd (Latin), CoPilot, ... and many others that no longer exist. People won't translate, it's physical work!
I'll translate it myself, no problem.

PS: When the new TTS text is done and when it is usable, the **promotion** of the application starts. Now, like this, the cripple cannot.

**Reply by zyphlar** [reply]
@ZeljkoBG I absolutely understand. Let me know if I can help with making the issue or organizing the files/etc. I just can't review the final result since I don't speak the language :)

One thing we will often do is use a script to Google Translate everything, and then you can just skim and review for obvious mistakes. Often that's good enough for 90% and it's just oddballs like "shop as a verb or noun?" that need extra care.

**Reply by ZeljkoBG** [reply]
@zyphlar will they add my translation https://github.com/user-attachments/files/17948072/strings.txt 
in the https://github.com/organicmaps/organicmaps/blob/master/data/strings/strings.txt ?

**Reply by zyphlar** [reply]
@ZeljkoBG we can open a pull request to do so however it will be best to at least attempt to translate more than just that file before merging. Would you like me to use Google Translate to begin the effort to translate all strings, and you can review it?

**Reply by ZeljkoBG** [reply]
I opened the topic as you advised me https://github.com/orgs/organicmaps/discussions/9858

There, give me advice on what to translate next. I would like to add s**tring.txt** to the new version of the application and when I finish translating the other files to go to the next version. If possible, please.

**Reply by zyphlar** [reply]
Ok thanks! Unfortunately it's not very good to only translate one file, load it into the application, and then begin working on the other files. That would be a lot of work to produce a version of the app that would be obviously broken (English) in many ways. 

I think the best way to proceed is to take your file and a bunch of auto-translated other files into a Pull Request, have you or any other Serbian speakers skim that PR for sanity, and _then_ add it to the mainline app for beta testing and further fixes.

I know it's less fun this way but it's hard to review a complete product like an app when there are major unfinished areas. Poking around the interface looking for Serbian strings when a bunch of the interface is still in English makes it hard to know what to make notes on and what will be fixed later (and then you have to review it twice?)

I'll try and make this translation and pull request soon, but anyone else is welcome to. The list of files to be translated is posted above, and yes all of them are required even (especially?) the store descriptions etc.

**Comment by ZeljkoBG** [comment]
@zyphlar 
Sorry, can we have any info, when will there be a new TTS?
